src/test/shader.js

```diff
@@ -0,0 +1,640 @@
+  if (value === gl.NO_ERROR) {
+    return "NO_ERROR";
+  }
+  for (var p in gl) {
+    if (gl[p] == value) {
+      return p;
+    }
+  }
+  return "0x" + value.toString(16);
+};
+
+function addSpan(contents, div) {
+  if (div == undefined) {
+    var divs = document.body.getElementsByClassName("testInfo");
+    var lastDiv = divs[divs.length - 1];
+    div = lastDiv;
+  }
+
+  var span = document.createElement("span");
+  div.appendChild(span);
+  span.innerHTML = contents + '<br />';
+}
+
+function addDiv(contents) {
+    var testInfoDiv = document.createElement("div");
+    document.body.appendChild(testInfoDiv);
+    testInfoDiv.setAttribute("class", "testInfo");
+
+    return testInfoDiv;
+}
+
+function logToConsole(msg) {
+  if (window.console)
+    window.console.log(msg);
+}
+
+function escapeHTML(text) {
+  return text.replace(/&/g, "&amp;").replace(/</g, "&lt;");
+}
+
+function testPassed(msg) {
+  addSpan('<span><span class="pass" style="color:green">PASS</span> ' + escapeHTML(msg) + '</span>');
+  logToConsole('PASS ' + msg);
+}
+
+function testFailed(msg) {
+  addSpan('<span><span class="fail" style="color:red">FAIL</span> ' + escapeHTML(msg) + '</span>');
+  logToConsole('FAIL ' + msg);
+}
+
+function glErrorShouldBe(gl, glErrors, opt_msg) {
+  if (!glErrors.length) {
+    glErrors = [glErrors];
+  }
+  opt_msg = opt_msg || "";
+  var err = gl.getError();
+  var ndx = glErrors.indexOf(err);
+  var errStrs = [];
+  for (var ii = 0; ii < glErrors.length; ++ii) {
+    errStrs.push(glEnumToString(gl, glErrors[ii]));
+  }
+  var expected = errStrs.join(" or ");
+  if (ndx < 0) {
+    var msg = "getError expected" + ((glErrors.length > 1) ? " one of: " : ": ");
+    testFailed(msg + expected +  ". Was " + glEnumToString(gl, err) + " : " + opt_msg);
+  } else {
+    //var msg = "getError was " + ((glErrors.length > 1) ? "one of: " : "expected value: ");
+    //testPassed(msg + expected + " : " + opt_msg);
+  }
+};
+
+// Special utility functions for df64 tests
+
+function df64ify(a) {
+  const a_hi = Math.fround(a);
+  const a_lo = a - Math.fround(a);
+  return new Float32Array([a_hi, a_lo]);
+}
+
+function getFloat64(upper = 256) {
+  return Math.random() * Math.pow(2.0, (Math.random() - 0.5) * upper);
+}
+
+function getVec4Float64() {
+  return [getFloat64(), getFloat64(), getFloat64(), getFloat64()]
+}
+
+function getMat4Float64() {
+  var result = [];
+  for (var i = 0; i < 16; i++)
+  {
+    result.push(getFloat64());
+  }
+  return result;
+}
+
+function initializeGL(canvas)
+{
+  const gl = createGLContext(canvas);
+  gl.viewport(0, 0, canvas.width, canvas.height);
+  gl.clearColor(0, 0, 0, 1);
+  gl.clearDepth(1);
+  gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
+
+  var fp_texture_support = gl.getExtension('OES_texture_float');
+  if (!fp_texture_support)
+  {
+    console.error("no floating point texture support!");
+  }
+  return gl;
+}
+
+function initializeTexTarget(gl)
+{
+  var framebuffer = gl.createFramebuffer();
```

Comment:
It would be really neat to do this with the luma Framebuffer/Texture classes. We should be able to write this with very few lines of code, and have it as an example of how to write shader tests. 
I'd be happy to look at that if you think it makes sense.